### PR TITLE
Remove all installers and install logic

### LINKED ISSUE #334: remove make_installer and all auto-install logic
We need to get out of the business of auto-installing things. It's too much scope. 

in the configs where we have  `make_installer` , 70-80% of the code ends up dealing with install logic. That isn't really the point of this repo , it's a distraction.

Note:

- **Documention** of common install steps **is valuable and appropriate for this repo**. E.g.:
  -  URL to download the server installer
  -  URL to upstream documentation explaining how to install the server 
  - brief instructions that user can copy/paste into their shell to install via common package managers such as apt-get/homebrew
- It could be valuable for `:checkhealth` to be the entrypoint that users can reach for, to detect common problems with LSP server installation. For that purpose some of the existing installer logic might be useful.

### PR DESCRIPTION
**Pull request by mjlbach**
Closes #334 #26 #92 #289 #359 #361 #421 #467

### CODE REVIEW

**Comment by mjlbach** [comment]
@teto This is going to take fairly exhaustive testing, and we'll probably have to update the documentation. It's mostly ready, and I think I removed all dead documentation. To be honest, I don't know how to install a lot of these language servers the "normal" way, since all of mine are packaged using nix.

**Comment by pretentious7** [comment]
:clap: :clap: :clap: 

I've been using basically that with typescript for a couple days, so can confirm that works.

File: lua/lspconfig/jsonls.lua
Comment (teto) [inline lua/lspconfig/jsonls.lua:21]: installed ?

File: lua/lspconfig/metals.lua
Comment (teto) [inline lua/lspconfig/metals.lua:33]: maybe reference that page ?

**Comment by teto** [comment]
thanks for tackling this, it needs to be done before the 0.5 release so that users dont start relying on these primitives.
Once this is merged, we need to reference it in https://github.com/neovim/neovim/wiki/Following-HEAD .

File: lua/lspconfig/metals.lua
Comment (mjlbach) [reply lua/lspconfig/metals.lua]: Done, and I added instruction. It's absurdly difficult to find documentation for how to manually install their language server, I had to look under the emacs instructions for a link, I copied the installation steps just to be safe).

**Comment by mjlbach** [comment]
If there are any remaining dicey documentation issues let me know. I think we should also fix whatever is going on with the print warning that cmd is undefined before merging (and hopefully more people chime in that the instructions are great/bad/insufficient. 

This will also break things for people, as cmd will need to be defined. So I imagine an influx of issues. I don't think there's a good intermediate measure.

**Comment by teto** [comment]
I think we should default `cmd` to the name of the binary and let the user put it in PATH. The good thing about this is that instead of a (lua) error (?), we can advise to run `:checkhealth nvim_lsp` and it will tell the user that the binary is not in PATH ? Ideally we need a much more helpful checkhealth for lsp https://github.com/neovim/neovim/pull/13161

**Comment by mjlbach** [comment]
> I think we should default cmd to the name of the binary and let the user put it in PATH.

I initially tried doing this, but many "commands" are simply the path to some random output of an npm build process. My strategy was mostly to keep the names of commands that point to actual binaries one would reasonably put on their path and remove those that were some random build output file. Also there's a slight issue in that windows binaries often will have a different extension. Right now without the error handling I put in, you'll get a nasty red error message if the language client fails to start the server.

File: lua/lspconfig/metals.lua
Comment (ckipp01) [inline lua/lspconfig/metals.lua:46]: I would change this as bootstrapping metals with `-Dmetals.client=emacs` sets some defaults that will cause Metals to not really work as expected. There actually is a very small section on using other vim LSP clients on the vim page here https://scalameta.org/metals/docs/editors/vim.html#using-an-alternative-lsp-client. However, that will be updated soon with nvim LSP specific stuff as well, since we're actually putting effort into having a nvim LSP focused extension for Metals users: https://github.com/scalameta/nvim-metals. This is mainly due to a  lot of defaults not working unless some extra things are offered, and we make heavy use of LSP extensions. Also an alternative to what you have here can I suggest the following?

```suggestion
See full instructions in the Metals documentation:

https://scalameta.org/metals/docs/editors/vim.html#using-an-alternative-lsp-client

Note: that if you're using [nvim-metals](https://github.com/scalameta/nvim-metals), that plugin fully handles the setup and installation of Metals, and you shouldn't set up Metals both with it and this plugin.

To install Metals, make sure to have [coursier](https://get-coursier.io/docs/cli-installation) installed, and once you do you can install the latest Metals with `cs install metals`. You can also manually bootstrap Metals with the following command.

```bash
cs bootstrap \
  --java-opt -Xss4m \
  --java-opt -Xms100m \
  org.scalameta:metals_2.12:<enter-version-here> \
  -r bintray:scalacenter/releases \
  -r sonatype:snapshots \
  -o /usr/local/bin/metals -f
```
```

**Comment by mjlbach** [comment]
@ckipp01 Thanks for the useful feedback! I've merged your changes. I don't use metals, and had a bit of a hard time finding the page you referenced, so I'm glad that is being worked on. I'm happy to nudge people to the nvim-metals plugin in the documentation; I don't think we can ever match the deep integration that server specific plugins have.

Let me know if there are any specific features you want that would make the plugin easier. I've recently filed several PRs with core improving 3.16 compliance.

**Comment by mjlbach** [comment]
@teto Unless I'm mistaken, I think the only bin_name we're missing now is elixir

**Comment by mjlbach** [comment]
@mortepau Do you have suggestions for improving the documentation regarding elixir install and default cmd without LspInstall?

**Comment by saadparwaiz1** [comment]
@mjlbach I'm not sure if this is helpful but I have been using the elixir language server with emacs's lsp-mode and the installation instruction are as simple picking up the zip file from github release and using language_server.sh on unix and language_server.bat on windows. These can be found [here](https://github.com/elixir-lsp/elixir-ls#detailed-installation-instructions)

**Comment by mjlbach** [comment]
@teto I think this is good for another look. I added the missing elixir/sumneko lua documentation, and added some documentation that had always been missing for OmniSharp.

**Comment by teto** [comment]
Is it possible to fix the CONFIG.md conflict. I will let @h-michael merge this when he feels like it, he is the most knowledgeable about this repo.

**Comment by mjlbach** [comment]
Fixed

File: lua/lspconfig.lua
Comment (teto) [inline lua/lspconfig.lua:24]: neovim 0.5 hasn't been released yet so in theory we could just get rid of this but aving the commands print a message like `print("deprecated, see https://github.com/neovim/neovim/wiki/Following-HEAD")  would avoid some pain for everyone (users and contributors on gitter). And we would remove it just before 0.5.

Comment (mjlbach) [reply lua/lspconfig.lua:24]: Added in a separate commit to make it easy to remove 🔪

File: lua/lspconfig/als.lua
Comment (teto) [inline lua/lspconfig/als.lua:22]: https://github.com/AdaCore/ada_language_server#Install

**Review by teto — APPROVED** [review]
I've grepped the source for LspInstall and there is one mention left in health.lua

**Comment by mjlbach** [comment]
I noticed that one, but I'm not sure what it means...

**Comment by teto** [comment]
git blame shows it was due to https://github.com/neovim/nvim-lspconfig/issues/248 . Seems like make_config  is not available prior to te setup call ? maybe keep the `if`  and just remove the comment about LspInstall for now. I think we can merge this today,

**Comment by mjlbach** [comment]
Done, I reworded it to reflect what I (think?) is the purpose.

**Comment by teto** [comment]
cool, can you mention the change on https://github.com/neovim/neovim/wiki/Following-HEAD ? feel free to merge when it's done

**Comment by mjlbach** [comment]
Done, and merging.

**Comment by Borwe** [comment]
Man I miss the auto LspInstall... was so easy.

**Comment by teto** [comment]
I've seen https://github.com/anott03/nvim-lspinstall popup in my github notifs. Dont know how good it is but may be a solution.

**Pull request merged**
